fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -0,0 +1,749 @@
+            ])
+            .to_string()
+    }
+}
+
+/// A structed of all parsed and validated values from the masked content bytes.
+#[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Hash, Serialize, Deserialize)]
+pub struct ParsedMaskedContent {
+    header: JWTHeader,
+    iss: String,
+    client_id: String,
+    hash: String,
+}
+
+/// Struct that represents a standard JWT header according to
+/// https://openid.net/specs/openid-connect-core-1_0.html
+#[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Hash, Serialize, Deserialize)]
+pub struct JWTHeader {
+    alg: String,
+    kid: String,
+    typ: String,
+}
+
+impl ParsedMaskedContent {
+    /// Parse the masked content bytes into a [struct ParsedMaskedContent].
+    /// Aux inputs (payload_start_index, payload_len, num_sha2_blocks) are
+    /// are used for validation and parsing.
+    pub fn new(
+        masked_content: &[u8],
+        payload_start_index: usize,
+        payload_len: usize,
+        num_sha2_blocks: usize,
+    ) -> Result<Self, FastCryptoError> {
+        // Verify the bytes after 64 * num_sha2_blocks should be all 0s.
+        if !masked_content[64 * num_sha2_blocks..]
+            .iter()
+            .all(|&x| x == 0)
+        {
+            return Err(FastCryptoError::GeneralError(
+                "Incorrect payload padding".to_string(),
+            ));
+        }
+
+        let masked_content_tmp = &masked_content[..64 * num_sha2_blocks];
+
+        // Verify the byte at payload start index is indeed b'.'.
+        if masked_content_tmp.get(payload_start_index - 1) != Some(&b'.') {
+            return Err(FastCryptoError::GeneralError(
+                "Incorrect payload index for separator".to_string(),
+            ));
+        }
+
+        let header = parse_and_validate_header(
+            masked_content_tmp
+                .get(0..payload_start_index - 1)
+                .ok_or_else(|| {
+                    FastCryptoError::GeneralError(
+                        "Invalid payload index to parse header".to_string(),
+                    )
+                })?,
+        )?;
+
+        // Parse the jwt length from the last 8 bytes of the masked content.
+        let jwt_length_bytes = masked_content_tmp
+            .get(masked_content_tmp.len() - 8..)
+            .ok_or_else(|| FastCryptoError::GeneralError("Invalid last 8 bytes".to_string()))?;
+        let jwt_length = calculate_value_from_bytearray(jwt_length_bytes);
+
+        // Verify the jwt length equals to 8*(payload_start_index + payload_len).
+        if jwt_length != 8 * (payload_start_index + payload_len) {
+            return Err(FastCryptoError::GeneralError(
+                "Incorrect jwt length".to_string(),
+            ));
+        }
+
+        // Parse sha2 pad into a bit array.
+        let sha_2_pad = bytearray_to_bits(&masked_content_tmp[payload_start_index + payload_len..]);
+
+        // Verify that the first bit of the bit array of sha2 pad is 1.
```

Comment:
Move code from lines 449-458 into a function titled `validate_sha2_pad`? The logic in `validate_zeros_count` can be a part of this.